### DISCUSSION TITLE
Need help to set Chrome Download Directory

### DISCUSSION THREAD

**Discussion by lewgaowei** [seleniumbase/SeleniumBase#2422] [Q&A] [answered]
**with normal selenium i can set the download directory by:**
prefs = {'download.default_directory': download_path}
chrome_options.add_experimental_option('prefs', prefs)

**How do i add the above to seleniumbase as per below:**

chromeProfileDirectory = f"{directory}/ChromeProfile/" + uniqueChromeProfile
self.driver = Driver(uc=True, incognito=False, user_data_dir=chromeProfileDirectory, )  ## SAVE CHROME PROFILE

**Comment by mdmintz** [comment, marked as answer]
Duplicate of https://github.com/seleniumbase/SeleniumBase/issues/1557#issuecomment-1287636878

--------

SeleniumBase creates the ``downloaded_files`` folder in the working directory where ``pytest`` is invoked. Any click-initiated downloads will go there. It's also used for any special files needed. It also holds lock files to prevent issues with multi-threading. The folder resets at the start of every new ``pytest`` run so that past test runs don't interfere with new ones. The folder is hard-coded there to prevent issues. And since the folder is reset at the start of new ``pytest`` runs, you wouldn't want to use any other existing folders for it.

There are lots of built-in test methods that are specially made for that folder, such as:

```python
self.get_downloads_folder()

self.get_browser_downloads_folder()

self.get_path_of_downloaded_file(file)

self.is_downloaded_file_present(file)

self.delete_downloaded_file_if_present(file)
# Duplicates: self.delete_downloaded_file(file)

self.assert_downloaded_file(file)

self.get_downloaded_files(regex=None, browser=False)

self.get_data_from_downloaded_file(file, timeout=None, browser=False)

self.assert_data_in_downloaded_file(data, file, timeout=None, browser=False)
```

If you need files in a different folder, use Python ``os``, ``sys``, and ``shutil`` libraries for copying the files you need into a different folder after they get downloaded to the ``downloaded_files`` folder.

There's also a `pytest` command-line option to archive existing downloads instead of deleting them:
https://github.com/seleniumbase/SeleniumBase/blob/570910c6f6e079cddd00c23e826e202c6489f39e/README.md?plain=1#L673

(A copy of those downloads will go into a folder in the `archived_files/` folder at the end of the `pytest` run.)

**Reply by andranikh89** [reply]
This works fine when we are in the single machine environment, but what if my setup involves Selenium Grid? I run the testcases from a Jenkins Node and the actual webdriver works on a different node from SeleniumGrid. So the script running the testcases seems to not see the files in selenium grid node. I configured a shared folder between these nodes, and if I can reconfigure the seleniumbase to download the files on the Grid node in the folder that I need, I can see them in Jenkins node.
at least this was my plan, but it doesn't work as I don't have a way to speicify the download directory.
What solution will you advice in this scenario?

**Reply by mdmintz** [reply]
If you have to test downloading files, then I wouldn't use the Grid for that. The Grid is great for controlling what happens inside a browser, but outside the browser (eg. file system) is a problem.

**Reply by andranikh89** [reply]
If I'd have a way to specify the download directory path with selenium base, it would not be a problem at all. SeleniumBase is a great library I use it for about 1 year now, and this was the first and only limitation I saw.
Anyways, I fixed the issue with a little hack by going other way around. I spotted what's the directory that SeleniumBase tries to download files, in the script that creates my nodes I make it to create this directory on each node and make it shared between my jenkins node and grid nodes. Not the most reliable approach as if SeleniumBase will decide to download in other directory, my logic will fail. So the best would be if you will give a way to customize this directory, if chrome driver provides this functionality, I don't see a reason to limit that functionality.
